component/alertrules.libsonnet

```diff
@@ -0,0 +1,123 @@
+  // whether its last occurrence in the input array was prefixed with ~ or
+  // not.
+  local val_state = std.foldl(
+    function(a, it) a + it,
+    [
+      { [realval(v)]: !std.startsWith(v, '~') }
+      for v in arr
+    ],
+    {}
+  );
+  // Return filtered array containing only elements whose last occurrence
+  // wasn't prefixed by ~.
+  std.filter(
+    function(val) val_state[val],
+    std.objectFields(val_state)
+  );
+
+// Keep only alerts from params.ignore_alerts for which the last
+// array entry wasn't prefixed with `~`.
+local user_ignore_alerts = render_array(params.ignore_alerts);
+
+// Upstream alerts to ignore
+local ignore_alerts = std.set(
+  // Add set of upstream alerts that should be ignored from processed value of
+  // `params.ignore_alerts`
+  user_ignore_alerts
+);
+
+// Alert rule patches.
+// Provide partial objects for alert rules that need to be tuned compared to
+// upstream. The keys in this object correspond to the `alert` field of the
+// rule for which the patch is intended.
+local patch_alerts = {
+  FluentdQueueLengthIncreasing: {
+    'for': '12h',
+  },
+};
+
+/* FROM HERE: should be provided as library function by
+ * rancher-/openshift4-monitoring */
+// We shouldn't be expected to care how rancher-/openshift4-monitoring
+// implement alert managmement and patching, instead we should be able to
+// reuse their functionality as a black box to make sure our alerts work
+// correctly in the environment into which we're deploying.
+
+local global_alert_params = inv.parameters.openshift4_monitoring.alerts;
+
+local filter_patch_rules(g) =
+  // combine our set of alerts to ignore with the monitoring component's
+  // set of ignoreNames.
+  local ignore_set = std.set(global_alert_params.ignoreNames + ignore_alerts);
+  g {
+    rules: std.map(
+      // Patch rules to make sure they match the requirements.
+      function(rule)
+        local rulepatch = com.getValueOrDefault(patch_alerts, rule.alert, {});
+        //        local runbook_url = runbook(rule.alert);
+        rule {
+          // Change alert names so we don't get multiple alerts with the same
+          // name, as the logging operator deploys its own copy of these
+          // rules.
+          alert: 'SYN_%s' % super.alert,
+          labels+: {
+            // ensure the alerts are not silenced on OCP4
+            // TODO: figure out how to ensure we don't get duplicate alerts on
+            // not-OCP4
+            syn: 'true',
+            // mark alert as belonging to openshift4-logging
+            // can be used for inhibition rules
+            syn_component: 'openshift4-logging',
+          },
+        } + rulepatch,
+      std.filter(
+        // Filter out unwanted rules
+        function(rule)
+          // only create duplicates of alert rules, we can use the recording
+          // rules which are deployed anyway when we enable monitoring on the
+          // CephCluster resource.
```

Comment:
Comment is just copy-pasted. We should check if the recording rules (if there are any) are actually deployed by the cluster-logging operator, and adjust the filter function if we need to deploy the recording rules ourselves.